drivers/timer/mchp_xec_rtos_timer.c

```diff
@@ -0,0 +1,297 @@
+ * in the future.
+ *
+ * Per comment below we don't clear RTMR pending interrupt.
+ * RTMR counter register is read-only and is loaded from the preload
+ * register by a 0->1 transition of the control register start bit.
+ * Writing a new value to preload only takes effect once the count
+ * register reaches 0.
+ * We have observed an issue. Sometimes the following sequence fails:
+ * Clear start bit, write new preload value, set start bit. The new
+ * preload value does not get copied to the count register. In fact,
+ * count goes to 0 and wraps. The work-around is to clear the block
+ * enable bit in the control register which clears all other bits and
+ * registers, write the new preload value, write control to start.
+ */
+void z_clock_set_timeout(s32_t n, bool idle)
+{
+	ARG_UNUSED(idle);
+
+	u32_t ccr, temp;
+	int   full_ticks;	/* number of complete ticks we'll wait */
+	u32_t full_cycles;	/* full_ticks represented as cycles */
+	u32_t partial_cycles;	/* number of cycles to first tick boundary */
+
+	if (n < 1) {
+		full_ticks = 0;
+	} else if ((n == K_FOREVER) || (n > MAX_TICKS)) {
+		full_ticks = MAX_TICKS - 1;
+	} else {
+		full_ticks = n - 1;
+	}
+
+	/*
+	 * RTMR frequency is fixed at 32KHz resulting in truncation errors.
+	 * Tune the denominator to take into account delay in caller and this
+	 * routine.
+	 */
+	full_cycles = (full_ticks * CYCLES_PER_TICK)
+		      + ((full_ticks * CPT_FRACT) / 1024UL);
+
+	/*
+	 * There's a wee race condition here. The timer may expire while
+	 * we're busy reprogramming it; an interrupt will be queued at the
+	 * NVIC and the ISR will be called too early, roughly right
+	 * after we unlock, and not because the count we just programmed has
+	 * counted down. We can detect this situation only by using one-shot
+	 * mode. The counter will be 0 for a "real" interrupt and non-zero
+	 * if we have restarted the timer here.
+	 */
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	ccr = RTMR_REGS->CNT;
+	total_cycles += (cached_icr - ccr);
+	partial_cycles = CYCLES_PER_TICK - (total_cycles % CYCLES_PER_TICK);
+	/* Adjust partial cycles by one. */
+	temp = full_cycles + partial_cycles + 1U;
+
+	/*
+	 * Write zero to control register causes HW to clear count register.
+	 * Write new preload register value, read-back, and write to read-only
+	 * count register adding small amount of delay before starting.
+	 */
+	RTMR_REGS->CTRL = 0;
+	RTMR_REGS->PRLD = temp;
+	RTMR_REGS->CNT = RTMR_REGS->PRLD;
+	RTMR_REGS->CTRL = TIMER_START_VAL;
+	cached_icr = temp;
+
+	k_spin_unlock(&lock, key);
+}
+
+/*
+ * Return the number of Zephyr ticks elapsed from last call to
+ * z_clock_announce in the ISR.
+ */
+u32_t z_clock_elapsed(void)
+{
+	u32_t ccr;
+	u32_t ticks;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	ccr = RTMR_REGS->CNT;
+	ticks = total_cycles - last_announcement;
+	ticks += cached_icr - ccr;
+	k_spin_unlock(&lock, key);
+	ticks /= CYCLES_PER_TICK;
```

Comment:
Style: you're reading a value in cycles and assigning it to a variable named "ticks", which you then convert at the end of the function.  Would be clearer if you named the variable "cyc" and then did a "`return cyc / CYCLES_PER_TICK;`"